source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -398,6 +419,44 @@ void DnsResolverImpl::AddrInfoPendingResolution::startResolutionImpl(int family)
       this);
 }
 
+DnsResolverImpl::AddrInfoPendingResolution::AvailableInterfaces
+DnsResolverImpl::AddrInfoPendingResolution::availableInterfaces() {
+  if (!Api::OsSysCallsSingleton::get().supportsGetifaddrs()) {
+    // Maintain no-op behavior if the system cannot provide interface information.
+    return {true, true};
+  }
+
+  Api::InterfaceAddressVector interface_addresses{};
+  const Api::SysCallIntResult rc = Api::OsSysCallsSingleton::get().getifaddrs(interface_addresses);
+  RELEASE_ASSERT(!rc.return_value_, fmt::format("getiffaddrs error: {}", rc.errno_));
+
+  DnsResolverImpl::AddrInfoPendingResolution::AvailableInterfaces available_interfaces{false,
+                                                                                       false};
+  for (const auto& interface_address : interface_addresses) {
+    if (Network::Utility::isLoopbackAddress(*interface_address.interface_addr_)) {
+      continue;
+    }
+
+    switch (interface_address.interface_addr_->ip()->version()) {
```

Comment:
Do we need to be concerned about interface_addr_->ip() returning nullptr? (I confess I'm a bit unclear how non-IP based addresses play with envoy)

Reply:
Good point. I didn't here because Envoy's getifaddrs impl only returns ip addresses. However, we do allow for alternate implementations. Updated to be defensive